src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/CachedExpressionCompiler.cs

```diff
@@ -107,17 +107,63 @@ private static class Compiler<TModel, TResult>
                 // by around one microsecond, so it's not worth it to complicate the logic here with
                 // an architecture check.
 
-                var memberExpression = expression.Body as MemberExpression;
-                if (memberExpression != null)
+                if (expression.Body is MemberExpression memberExpression)
                 {
-                    if (memberExpression.Expression == expression.Parameters[0] || memberExpression.Expression == null)
+                    if (memberExpression.Expression == null)
                     {
-                        // model => model.Member or model => StaticMember
+                        // model => model => StaticMember
                         return _simpleMemberAccessCache.GetOrAdd(memberExpression.Member, _ => expression.Compile());
                     }
 
-                    var constantExpression = memberExpression.Expression as ConstantExpression;
-                    if (constantExpression != null)
+                    if (memberExpression.Expression == expression.Parameters[0])
+                    {
+                        // model => model.Member
+                        if (memberExpression.Member.DeclaringType.IsValueType &&
+                            Nullable.GetUnderlyingType(memberExpression.Type) == null)
+                        {
+                            // struct.Member
+                            return _simpleMemberAccessCache.GetOrAdd(memberExpression.Member, _ => expression.Compile());
```

Comment:
Since this stores lambdas, does this build a chain of lambdas (rather than compiling a single composite lambda)?